### DISCUSSION TITLE
Problem getting i2s mic to output to Analog (ESP32 pin 25/26)

### DISCUSSION THREAD

**Discussion by ShadGates** [pschatzmann/arduino-audio-tools#132] [Q&A] [answered]
Seems like this should be simple, following the basic pattern of the existing examples

I've tested the setup of the analog output via the streams-memory-mp3-analog and streams-generator-analog examples, They play through the speaker just fine.
I've tested the INMP441 via the streams-i2s-webserver_wav example, and it plays sound through the browser as it should.  I've also used a second INMP441, it also works with the webserver example, but fails with this sketch.

**Here's the debug output**
```
INMP441 i2s mic to Analog via onboard DAC to TDA7052 amp
starting Analog output...
[D] AnalogAudioESP32.h : 168 - defaultConfig
[I] AnalogAudioESP32.h : 70 - I2S_MODE_DAC_BUILT_IN
[I] AnalogAudioESP32.h : 195 - bool audio_tools::AnalogAudioStream::begin(audio_tools::AnalogConfig)
[I] AudioTypes.h : 38 - sample_rate: 44100
[I] AudioTypes.h : 39 - channels: 2
[I] AudioTypes.h : 40 - bits_per_sample: 16
[I] AnalogAudioESP32.h : 90 - analog left output pin: 25
[I] AnalogAudioESP32.h : 91 - analog right output pin: 26
[I] AnalogAudioESP32.h : 263 - TX_MODE
starting I2S mic...
[D] I2SStream.h : 92 - void audio_tools::I2SStream::begin(audio_tools::I2SConfig)
[D] I2SESP32.h : 38 - bool audio_tools::I2SBase::begin(audio_tools::I2SConfig)
[D] I2SESP32.h : 133 - bool audio_tools::I2SBase::begin(audio_tools::I2SConfig, int, int)
[I] I2SConfig.h : 89 - rx/tx mode: RX
[I] I2SConfig.h : 90 - port_no: 0
[I] I2SConfig.h : 91 - is_master: Master
[I] I2SConfig.h : 92 - sample rate: 44100
[I] I2SConfig.h : 93 - bits per sample: 16
[I] I2SConfig.h : 94 - number of channels: 2
[I] I2SConfig.h : 95 - i2s_format: I2S_STD_FORMAT
[I] I2SConfig.h : 97 - auto_clear:0
[I] I2SConfig.h : 104 - buffer_count:10
[I] I2SConfig.h : 105 - buffer_size:512
[I] I2SConfig.h : 109 - pin_bck: 14
[I] I2SConfig.h : 110 - pin_ws: 15
[I] I2SConfig.h : 111 - pin_data: 32
[D] I2SESP32.h : 161 - i2s_driver_install
[D] I2SESP32.h : 178 - i2s_set_pin
[D] I2SESP32.h : 189 - i2s_zero_dma_buffer
[D] I2SESP32.h : 193 - begin - started
started...
[D] AudioCopy.h : 91 - size_t audio_tools::StreamCopyT<T>::copy() [with T = unsigned char; size_t = unsigned int]
E (361) I2S: /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/driver/i2s.c:1352 (i2s_read):rx NULL
[E] I2SESP32.h : 93 - size_t audio_tools::I2SBase::readBytes(void*, size_t)
[I] AudioCopy.h : 121 - StreamCopy::copy 1024 -> 0 -> 0 bytes - in 0 hops
```
**As soon as it starts the copier.copy(), it starts throwing the above errors on every call.**

**The sketch...**
```
#include "AudioTools.h"

typedef int16_t sound_t;
uint16_t sample_rate=44100;
uint8_t channels = 2;

AnalogAudioStream out; 
I2SStream i2sStream;
StreamCopy copier(out, i2sStream); 

void setup(void) {  
  Serial.begin(115200);
  
  AudioLogger::instance().begin(Serial, AudioLogger::Debug);
  delay(200); // give logger time to initialize so it doesn't skip entries.
  
  Serial.println("   ");
  Serial.println("INMP441 i2s mic to Analog via onboard DAC to TDA7052 amp");
  Serial.println("starting Analog output...");
  auto configOut = out.defaultConfig(TX_MODE);
  configOut.sample_rate = sample_rate; 
  configOut.channels = channels;
  configOut.bits_per_sample = 16;
  out.begin(configOut);

  Serial.println("starting I2S mic...");
  auto config = i2sStream.defaultConfig(RX_MODE);
  config.i2s_format = I2S_STD_FORMAT; 
  config.sample_rate = sample_rate;
  config.channels = channels;
  config.bits_per_sample = 16;
  i2sStream.begin(config);

  Serial.println("started...");
}

void loop() {
  copier.copy();
}
```

I'm not seeing where I've gone off the path.  
I've tried sample_rate of 22050, like the webserver example, and 44100, as other examples use, same error condition with either value.

Is there a way to check the status of the i2sStream object to make sure it's in a valid state for the copier to use it?  

Any other debug tips?

Any help appreciated, Thanks!

**Comment by pschatzmann** [comment, marked as answer]
The AnalogAudioStream and I2SStream use the I2S functionality of the ESP32.  The ESP32 has 2 I2S ports and if I remember right the analog functionality needs to be on port 0, which is used by default if you do not specify anything.

So if you use two ports you need to specify this in the config: 
```
 config.port = 1; 
```
should make the difference...

**Reply by ShadGates** [reply]
Thanks for the clue about ports and analog requiring 0.

The config property is "port_no", and once I set it to 1 on the i2s config, and zero (the default) on the analog stream, the error went away.   

Now just have to figure out why not actually getting any sound out of it.. 

Thinking it was just a collision, and that the affinity may be reversed, I tried 1 for analog, 0 for i2s, and that caused a boot-loop on the ESP32, so your recollection was correct that Analog only likes port_no = 0.

Any other troubleshooting tips/suggestions on debugging other likely causes of data apparently not copying over for output?
